### PR TITLE
Migrate to Django 4

### PR DESCRIPTION
**Pull request by christophmeissner**
Uncluttered (rebased onto `develop`, cherry-picked and partially squashed) branch for modernizing the project and migrate to Django 4.

### CODE REVIEW

**Comment by christophmeissner** [comment]
I think, we can start reviewing this.

File: requirements/production.txt [hunk 1]
```diff
@@ -1,2 +1,3 @@
 -r base.txt
+
 MySQL-python==1.2.5
```
Comment (pitpalme) [inline requirements/production.txt:3]: Can we *please* talk about "which production database?"? o.O

File: requirements/base.txt [hunk 2]
```diff
@@ -1,32 +1,11 @@
...
+django==4.0.3
+django-ckeditor==6.2.0
+django-compressor==3.1
+django-debug-toolbar==3.2.4
+django-extensions==3.1.5
+django-logentry-admin==1.1.0
+django-registration-redux==2.10
+git+https://github.com/Doca/django-ajax.git#egg=djangoajax
+
+factory-boy==3.2.1
```
Comment (pitpalme) [inline requirements/base.txt:10]: Is it really "base" or only "dev"?

File: requirements/dev.txt [hunk 3]
```diff
@@ -1,8 +1,17 @@
 -r base.txt
 -r tests.txt
 
+# debugging
+ipdb==0.13.9
+ipython==8.1.1
```
Comment (pitpalme) [inline requirements/dev.txt:6]: Can we please keep ipython in "base"? It's useful for administration of a running instance.

File: scheduler/__init__.py [hunk 4]
```diff
@@ -1,2 +1 @@
 # coding=utf-8
-default_app_config = 'scheduler.apps.SchedulerConfig'
```
Comment (pitpalme) [inline scheduler/__init__.py:2]: Why removing it? It defined translated verbose name, using e. g. in admin backend.

File: scheduler/managers.py [hunk 5]
```diff
@@ -18,9 +18,8 @@ def on_shiftdate(self, shiftdate):
             shiftdate. That means shifts that intersect with the day of
             shiftdate.
         """
-        next_day = datetime.combine(shiftdate + timedelta(days=1), time.min)
         return self.filter(ending_time__gte=shiftdate,
-                           starting_time__lt=next_day)
+                           starting_time__lt=shiftdate + timedelta(days=1))
```
Comment (pitpalme) [inline scheduler/managers.py:22]: Old code set's time of day to zero (`combine` + `time.min`), to filter shifts starting "before midnight".
New code seems to filter for anything starting less than 24 hours after reference.
`shiftdate` currently is assembled only from "year", "month" and "day", but there's no guaranty, no-one will be using "on_shiftdate" with some `datetime` having "time != 0".
IMHO we should explicitly make sure, time is zero.

File: scheduletemplates/models.py [hunk 6]
```diff
@@ -71,18 +71,16 @@ class Meta:
...
+        end = datetime.combine(today + timedelta(days=self.days), self.ending_time)
         duration = end - start
         return duration
 
     @property
     def localized_display_ending_time(self):
         days = self.days if self.ending_time > time.min else 0
-        days_fmt = ungettext_lazy(u'the next day',
-                                  u'after {number_of_days} days', days)
+        days_fmt = _(u'after {number_of_days} days'.format(number_of_days=days))
```
Comment (pitpalme) [inline scheduletemplates/models.py:83]: Why trashing pluralized translation?

**Review by pitpalme — CHANGES_REQUESTED** [review]
Questions and change request inlined.

File: scheduletemplates/models.py (hunk 6 above)
Comment (Doca) [reply scheduletemplates/models.py:83]: ugettext_lazy vs ungettext_lazy changed its signiture and I felt that this is not a big miss ;-)

File: scheduler/managers.py (hunk 5 above)
Comment (christophmeissner) [reply scheduler/managers.py:22]: Good point!

File: tests/registration/test_registration.py [hunk 7]
```diff
@@ -126,7 +129,7 @@ def test_passwords_dont_match(self):
 
         # TODO: why is this a non-field error? Shouldn't it be a error on the
         # second password field?
-        self.assertFormError(response, "form", "password2", "The two password fields didn't match.")
+        self.assertFormError(response, "form", "password2", "The two password fields didn’t match.")
```
[the remark this answers is not in the document]
Comment (christophmeissner) [reply tests/registration/test_registration.py:132]: It was changed in Django https://github.com/django/django/pull/11281

File: volunteer_planner/settings/production.py [hunk 8]
```diff
@@ -50,13 +50,5 @@
...
-    INSTALLED_APPS += (
-        'debug_toolbar',
-    )
 
-    MIDDLEWARE = [
-        'debug_toolbar.middleware.DebugToolbarMiddleware',
-    ] + MIDDLEWARE
-
-    INTERNAL_IPS = ['127.0.0.1', '172.20.0.1']
+INTERNAL_IPS = ['127.0.0.1', '172.20.0.1']
```
[the remark this answers is not in the document]
Comment (christophmeissner) [reply volunteer_planner/settings/production.py:54]: This line was not changed in the PR and is not specific to Django 4. Let's address it separately.

File: scheduler/__init__.py (hunk 4 above)
Comment (christophmeissner) [reply scheduler/__init__.py:2]: See commit message: 

> Remove default_app_config
> 
> RemovedInDjango41Warning:
> Django now detects this configuration automatically. You can remove default_app_config.
>

File: requirements/base.txt (hunk 2 above)
Comment (christophmeissner) [reply requirements/base.txt:10]: `factory-boy` is used in the `create_dummy_data` management command (and tests, but that's ). Currently all management commands are deployed to production... this makes this a requirement for all environments, I guess? However, we could remove it from base and move it to dev, in order to preven accidential execution of the command on a production environment. 

In the future: to make specific commands unavailable in a certain environment, we could collecte debugging and development commands in a new app (eg. `devtools`), and add it to `INSTALLED_APPS` only in non-production settings.

File: requirements/dev.txt (hunk 3 above)
Comment (christophmeissner) [reply requirements/dev.txt]: Done.

File: organizations/__init__.py [hunk 9]
```diff
@@ -1,2 +1 @@
 # coding=utf-8
-default_app_config = 'organizations.apps.OrganizationsConfig'
```
[the remark this answers is not in the document]
Comment (christophmeissner) [reply organizations/__init__.py:2]: see https://github.com/coders4help/volunteer_planner/pull/488#discussion_r834622773

File: requirements/production.txt (hunk 1 above)
Comment (christophmeissner) [reply requirements/production.txt:3]: I think, it's not specific to the update to Django 4. Let's address it separately.

**Review by pitpalme — APPROVED** [review]

**Comment by christophmeissner** [comment]
🎉🥳🍻 Good job, everybody! Congratulations!
